app/core/src/main/java/com/fsck/k9/cache/ExpiringCache.kt

```diff
@@ -0,0 +1,43 @@
+package com.fsck.k9.cache
+
+internal class ExpiringCache<KEY : Any, VALUE : Any>(
+    private val delegateCache: Cache<KEY, VALUE> = InMemoryCache(),
+    private var lastClearTime: Long = nowInMillis(),
+    private val cacheTimeValidity: Long = CACHE_TIME_VALIDITY_IN_MILLIS
+) : Cache<KEY, VALUE> {
+
+    override fun get(key: KEY): VALUE? {
+        recycle()
+        return delegateCache[key]
+    }
+
+    override fun set(key: KEY, value: VALUE) {
+        recycle()
+        delegateCache[key] = value
+    }
+
+    override fun hasKey(key: KEY): Boolean {
+        recycle()
+        return delegateCache.hasKey(key)
+    }
+
+    override fun clear() {
+        lastClearTime = nowInMillis()
+        delegateCache.clear()
+    }
+
+    private fun recycle() {
+        if (isExpired()) {
+            clear()
+        }
+    }
+
+    private fun isExpired(): Boolean {
+        return (nowInMillis() - lastClearTime) >= cacheTimeValidity
+    }
+
+    private companion object {
+        const val CACHE_TIME_VALIDITY_IN_MILLIS = 30_000L
+        fun nowInMillis(): Long = System.currentTimeMillis()
```

Comment:
Please use [`Clock`](https://github.com/thundernest/k-9/blob/d2e8dbb3ca635aa6b2563f0f222646ca76acba8b/app/core/src/main/java/com/fsck/k9/Clock.kt) to be able to inject values for testing.